File: misc/aooswd/ex/strategy/strategy.cpp
class Composition
{
public:
	Composition (Compositor*);
	void Repair ();
private:
	Compositor* _compositor;
	Component* _components;
	int _componentCount;
	int _lineWidth;
	int* _lineBreaks;
	int _lineCount;
};

class Compositor
{
public:
	virtual int Compose (Coord natural[], Coord stretch[], Coord shrink[], int componentCount, int lineWidth, int breaks[]) = 0;
protected:
	Compositor ();
};

void Composition::Repair ()
{
	Coord* natural;
	Coord* stretchability;
	Coord* shrinkability;
	int componentCount;
	int* breaks;
	// fill the arrays
	int breakCount;
	breakCount = _compositor -> Compose (natural, stretchability, shrinkability, componentCount, _lineWidth, breaks);
	// layout components
}

class SimpleCompositor : public Compositor
{
public:
	SimpleCompositor ();
	virtual int Compose (Coord natural[], Coord stretch[], Coord shrink[], int componentCount, int lineWidth, int breaks[]);
};

Composition* quick = new Composition (new SimpleCompositor);

File: misc/aooswd/ex/strategy/strategy.java
public class Composition
{
	public Composition (Compositor);
	public void Repair ()
	{
		Coord natural;
		Coord stretchability;
		Coord shrinkability;
		int componentCount;
		int breaks[];
		// fill the arrays
		int breakCount;
		breakCount = _compositor.Compose (natural, stretchability, shrinkability, componentCount, _lineWidth, breaks);
		// layout components
	}
	private Compositor _compositor;
	private Component _components;
	private int _componentCount;
	private int _lineWidth;
	private Vector _lineBreaks;
	private int _lineCount;
}

abstract public class Compositor
{
	abstract public int Compose (Coord natural[], Coord stretch[], Coord shrink[], int componentCount, int lineWidth, int breaks[]);
	protected Compositor ();
};

public class SimpleCompositor extends Compositor
{
	public SimpleCompositor ();
	int Compose (Coord natural[], Coord stretch[], Coord shrink[], int componentCount, int lineWidth, int breaks[]);
};

Composition quick = new Composition (new SimpleCompositor());